File: apps/mobile_flutter/lib/src/location/location_adapter.dart
import 'dart:async';

import 'package:geolocator/geolocator.dart';

import '../domain/types.dart';

/// Обёртка над geolocator. Скрывает платформенные детали от UI и state.
/// В Phase 1 заменяется полноценным `LocationAdapter` интерфейсом
/// (см. ТЗ §3.3 sensor-agnostic).
class LocationAdapter {
  StreamSubscription<Position>? _subscription;

  Future<void> start(void Function(RawPoint) listener) async {
    _subscription = Geolocator.getPositionStream(
      locationSettings: const LocationSettings(
        accuracy: LocationAccuracy.bestForNavigation,
        distanceFilter: 5,
      ),
    ).listen((position) {
      listener(_toRawPoint(position));
    });
  }

  Future<void> stop() async {
    await _subscription?.cancel();
    _subscription = null;
  }

  bool get isRunning => _subscription != null;
}

RawPoint _toRawPoint(Position p) => RawPoint(
  timestamp: p.timestamp.millisecondsSinceEpoch,
  latitude: p.latitude,
  longitude: p.longitude,
  altitude: p.altitude,
  accuracy: p.accuracy,
  speed: p.speed,
  heading: p.heading,
);

final LocationAdapter locationAdapter = LocationAdapter();

File: apps/mobile-rn/src/location/locationAdapter.ts
import * as Location from 'expo-location';
import type { RawPoint } from '../domain/types';

export type LocationListener = (point: RawPoint) => void;

/**
 * Обёртка над expo-location. Скрывает платформенные детали от UI и state.
 * В Phase 1 заменяется полноценным `LocationAdapter` интерфейсом
 * (см. ТЗ §3.3 sensor-agnostic).
 */
export class LocationAdapter {
  private subscription: Location.LocationSubscription | null = null;

  async start(listener: LocationListener): Promise<void> {
    this.subscription = await Location.watchPositionAsync(
      {
        accuracy: Location.Accuracy.BestForNavigation,
        distanceInterval: 5,
        timeInterval: 1000,
      },
      (location) => {
        listener(toRawPoint(location));
      },
    );
  }

  stop(): void {
    this.subscription?.remove();
    this.subscription = null;
  }

  isRunning(): boolean {
    return this.subscription !== null;
  }
}

function toRawPoint(location: Location.LocationObject): RawPoint {
  const { coords, timestamp } = location;
  return {
    timestamp,
    latitude: coords.latitude,
    longitude: coords.longitude,
    altitude: coords.altitude,
    accuracy: coords.accuracy,
    speed: coords.speed,
    heading: coords.heading,
  };
}

export const locationAdapter = new LocationAdapter();
